In [59]:
import pandas as pd
import numpy as np

In [60]:
df = pd.read_csv(r"C:\Users\Junayed\nyc-311-resolution-time-analysis\data\clean\311_jan_week1_clean.csv")

In [61]:
df.shape

(85122, 11)

### Analyze Full Frequency Distribution of Resolution Turnaround

Evaluate the value count distribution of `Resolution_Days` across all 389 unique duration lengths:

* **High Front-End Concentration:** Over $65\%$ of cases (**55,687**) are resolved on the same day (`0` days), and over $79\%$ are completed within 24–48 hours (`0`–`1` days: **67,888** tickets).
* **Rapid Drop-Off:** Resolution counts decay sharply from Day 0 down through Day 4 (**917** tickets), illustrating standard operational throughput for high-volume routine complaints.
* **Extreme Tail Dispersion:** The existence of 389 distinct day counts with single-ticket outliers stretching up to several hundred days (e.g., 222, 285, 336, 503, 582) visually confirms a heavily right-skewed distribution driven by delayed administrative closures or complex field investigations.

In [62]:
df["Resolution_Days"].value_counts()

Resolution_Days
0      55687
1      12201
2       5024
3       2261
4        917
       ...  
222        1
582        1
336        1
285        1
503        1
Name: count, Length: 389, dtype: int64

### Inspect Top 10 Most Frequent Resolution Durations

Examine the highest-frequency turnaround times to understand operational velocity across typical service requests:

* **Dominant Velocity:** Turnaround times of $0$ to $2$ days comprise the vast majority of the top tiers, led heavily by same-day closures (**55,687** tickets) and 1-day completions (**12,201** tickets).
* **Decay Trajectory:** Volume decreases monotonically from Day 0 through Day 7, reflecting standard resolution curves where simpler operational tasks are cleared early in the ticket lifecycle.
* **Slight Non-Linearity (Days 8–9):** Day 9 (**369** tickets) marginally outpaces Day 8 (**317** tickets), likely reflecting post-weekend administrative batch closures or scheduled review cycles.

In [63]:
df["Resolution_Days"].value_counts().head(10)

Resolution_Days
0    55687
1    12201
2     5024
3     2261
4      917
5      594
6      567
7      509
9      369
8      317
Name: count, dtype: int64

In [64]:
df["Resolution_Days"].value_counts().sort_index(ascending=False).head(10)

Resolution_Days
622    1
621    2
620    1
619    1
600    1
592    1
591    1
582    1
581    1
576    1
Name: count, dtype: int64

### Analyze Complaint Type Frequency Distribution

Evaluate the overall volume and diversity of service request categories across all 133 unique complaint classifications:

* **Primary Drivers:** The top three categories alone account for the vast majority of intake:
  * **Noise - Residential:** Dominates the dataset with **30,224** records ($>35\%$ of total complaints).
  * **Heat/Hot Water:** Represents the second largest share with **14,711** records, typical for seasonal winter demand in early January.
  * **Illegal Parking:** Logs **9,745** records, making quality-of-life and curb-space enforcement top municipal priorities.
* **Secondary Volume:** Issues like `Blocked Driveway` (**3,297**) and `Street Condition` (**1,310**) represent moderate recurring city service needs.
* **Long Tail Distribution:** The remaining categories taper off sharply into single-incident specialized complaints (e.g., `Scaffold Safety`, `Pet Shop`, `Building Condition`), illustrating broad operational scope across NYC agencies.

In [65]:
df["Complaint_Type"].value_counts()

Complaint_Type
Noise - Residential               30224
Heat/Hot Water                    14711
Illegal Parking                    9745
Blocked Driveway                   3297
Street Condition                   1310
                                  ...  
Posting Advertisement                 1
Pet Shop                              1
Building Condition                    1
Scaffold Safety                       1
Institution Disposal Complaint        1
Name: count, Length: 133, dtype: int64

### Inspect Top 10 Most Frequent Complaint Types

Examine the ten highest-volume service request categories to identify the primary municipal drivers across New York City:

* **Dominant Concerns:** Quality-of-life and residential comfort issues dominate citywide intake:
  * **Residential Living:** `Noise - Residential` (**30,224**) and `Heat/Hot Water` (**14,711**) alone account for over $52\%$ of all closed requests.
  * **Curb Management & Traffic:** `Illegal Parking` (**9,745**), `Blocked Driveway` (**3,297**), and `Abandoned Vehicle` (**1,195**) represent major street enforcement workloads.
* **Environmental & Infrastructure:** `Street Condition` (**1,310**), `Unsanitary Condition` (**1,273**), and `Water System` (**954**) reflect essential neighborhood infrastructure and sanitation maintenance.
* **Secondary Noise Channels:** Public and commercial noise complaints (`Noise - Commercial`: **1,058**; `Noise - Street/Sidewalk`: **1,054**) add further burden to municipal sound monitoring beyond residential properties.

In [66]:
df["Complaint_Type"].value_counts().head(10)

Complaint_Type
Noise - Residential        30224
Heat/Hot Water             14711
Illegal Parking             9745
Blocked Driveway            3297
Street Condition            1310
Unsanitary Condition        1273
Abandoned Vehicle           1195
Noise - Commercial          1058
Noise - Street/Sidewalk     1054
Water System                 954
Name: count, dtype: int64

### Final Geographic Distribution by Borough (Post-Cleaning)

Evaluate the final distribution of closed service requests across all five NYC boroughs following outlier removal, missing-value filtering, and deduplication:

* **Cohort Integrity:** The post-cleaning volume confirms **85,122** verified complaints fully attributed across the five boroughs with zero unspecified entries.
* **Volume Distribution:**
  * **Bronx:** Remains the primary source of service requests at **37,641** cases ($44.2\%$ of total volume).
  * **Brooklyn:** Represents the second-highest concentration with **18,859** complaints ($22.2\%$).
  * **Queens & Manhattan:** Account for **14,306** ($16.8\%$) and **12,243** ($14.4\%$) records, respectively.
  * **Staten Island:** Accounts for the smallest operational volume at **2,073** requests ($2.4\%$).
* **Consistency Check:** Proportional representation across boroughs remains stable compared to the initial raw intake, confirming that data hygiene filters did not introduce regional sampling bias.

In [67]:
df["Borough"].value_counts()

Borough
Bronx            37641
Brooklyn         18859
Queens           14306
Manhattan        12243
Staten Island     2073
Name: count, dtype: int64

### Analyze Service Request Volume by Responding Agency

Evaluate operational workload distribution across municipal departments handling closed 311 tickets:

* **Primary Responders:** Two agencies absorb over $80\%$ of total service request intake:
  * **NYPD (Police Department):** Leads citywide volume with **49,047** tickets ($57.6\%$), handling quality-of-life enforcement including residential/commercial noise, blocked driveways, and illegal parking.
  * **HPD (Housing Preservation & Development):** Manages **19,787** requests ($23.2\%$), predominantly addressing residential heating, hot water, and unsanitary building conditions.
* **Secondary Responders:**
  * **DSNY (Sanitation):** **4,674** tickets covering illegal dumping and collection concerns.
  * **DOT (Transportation):** **3,472** tickets addressing street, sidewalk, and traffic condition repairs.
  * **DEP (Environmental Protection):** **2,766** complaints for water system, sewer, and environmental noise issues.
* **Specialized Long Tail:** Specialized departments—including **DOB** (Buildings: 1,558), **DOHMH** (Health: 1,237), **DPR** (Parks: 898), **TLC** (Taxi & Limousine: 526), **EDC**, **DHS**, **DCWP**, **DOE**, and **OTI** (6 tickets)—manage low-volume, jurisdiction-specific cases.

In [68]:
df["Agency"].value_counts()

Agency
NYPD     49047
HPD      19787
DSNY      4674
DOT       3472
DEP       2766
DOB       1558
DOHMH     1237
DPR        898
TLC        526
EDC        452
DHS        355
DCWP       291
DOE         53
OTI          6
Name: count, dtype: int64

### Identify the Top 20 Slowest Service Request Categories

Rank complaint types by median turnaround duration (`.sort_values(ascending=False)`) to pinpoint operational bottlenecks and long-cycle municipal services:

* **Forestry & Tree Work (DPR):** Severe delays characterize arboricultural services—`Uprooted Stump` leads citywide resolution times at a median of **175.5 days**, followed by `New Tree Request` at **147.0 days** and `Root/Sewer/Sidewalk Condition` at **49.0 days**, reflecting seasonal planting schedules and specialized contractor dispatch.
* **Complex Environmental & Administrative Reviews:** 
  * `Noise - Helicopter` requires **138.0 days**, as aviation noise involves cross-jurisdictional reviews with the FAA and EDC.
  * Specialized building audits under `Special Projects Inspection Team (Spit)` require **101.0 days**.
* **Regulatory Compliance & Licensing (TLC & DOHMH):** Consumer regulatory matters take several months—`Green Taxi Complaint` (**96.5 days**), `For Hire Vehicle Complaint` (**79.0 days**), `Taxi Complaint` (**76.5 days**), and `Food Establishment` inspections (**60.0 days**) involve evidence audits, driver hearings, and scheduled physical site visits.
* **Property & Facility Conditions:** Cases like `Lot Condition` (**93.0 days**) and `Building/Use` (**63.0 days**) reflect legal notice periods, property owner remediation windows, and repeat inspection requirements.

In [69]:
df.groupby("Complaint_Type")["Resolution_Days"].median().sort_values(ascending=False).head(20)

Complaint_Type
Uprooted Stump                             175.5
New Tree Request                           147.0
Noise - Helicopter                         138.0
Special Projects Inspection Team (Spit)    101.0
Green Taxi Complaint                        96.5
Lot Condition                               93.0
For Hire Vehicle Complaint                  79.0
Taxi Complaint                              76.5
Bench                                       69.5
Building/Use                                63.0
Food Establishment                          60.0
Non-Residential Heat                        60.0
Smoking Or Vaping                           60.0
School Maintenance                          50.0
Root/Sewer/Sidewalk Condition               49.0
Graffiti                                    34.0
General                                     24.0
Appliance                                   22.0
Outside Building                            21.5
Unsanitary Condition                        20.0
Name:

### Identify the Fastest Service Request Categories (Same-Day Closures)

Inspect complaint categories sorted ascendingly by median resolution time to understand which issues receive immediate operational closure:

* **Instant Dispatch & Urgent Response (Median = 0.0 Days):** A substantial cluster of service request categories consistently close on the same calendar day they are created.
* **Public Safety & Police Intervention (NYPD):** Quality-of-life and acute public enforcement categories—including `Blocked Driveway`, `Abandoned Vehicle`, `Drinking`, `Drug Activity`, `Disorderly Youth`, and `Illegal Fireworks`—are dispatched directly to local precincts and resolved within hours.
* **Hazardous Environmental Health (DEP / DOHMH):** Immediate-response public health risks such as `Hazardous Materials` and `Asbestos` exhibit same-day median turnaround due to mandated emergency response protocols.
* **Operational Implication:** The heavy concentration of zero-day medians demonstrates why the citywide median resolution time remains $0.0$ days despite multi-month outliers in forestry and licensing.

In [70]:
df.groupby("Complaint_Type")["Resolution_Days"].median().sort_values(ascending=True).head(20)

Complaint_Type
Abandoned Vehicle                0.0
Animal In A Park                 0.0
Asbestos                         0.0
Animal-Abuse                     0.0
Blocked Driveway                 0.0
Building Condition               0.0
Bike/Roller/Skate Chronic        0.0
Best/Site Safety                 0.0
Derelict Vehicles                0.0
Drinking                         0.0
Disorderly Youth                 0.0
Dirty Condition                  0.0
Commercial Disposal Complaint    0.0
Bus Stop Shelter Placement       0.0
Encampment                       0.0
Drug Activity                    0.0
Hazardous Materials              0.0
Illegal Fireworks                0.0
Illegal Animal Sold              0.0
Illegal Animal Kept As Pet       0.0
Name: Resolution_Days, dtype: float64

### Quantify Unique Complaint Categories

Compute the number of distinct complaint types present in the cleaned dataset using `.nunique()`:

* **Category Count:** Exactly **133 unique complaint types** are represented across the 85,122 validated service requests.

In [71]:
complaints_type = df["Complaint_Type"].nunique()
print(complaints_type)

133


### Tabulate Complaint Category Volumes (`complaint_dist`)

Convert the frequency series of `Complaint_Type` into a structured summary table using `.value_counts().reset_index()`:

* **Structure:** Produces a 133-row reference table linking each complaint type directly to its absolute ticket volume (`Ticket_Count`), sorted in descending order of frequency.
* **Volume Distribution Highlights:**
  * **Head of Distribution:** Dominated by quality-of-life and residential comfort issues, led by `Noise - Residential` (**30,224**) and `Heat/Hot Water` (**14,711**).
  * **Tail of Distribution:** Displays multiple low-incidence categories logging exactly **1** occurrence (e.g., `Posting Advertisement`, `Pet Shop`, `Building Condition`, `Scaffold Safety`), demonstrating the long-tail operational profile of 311 intake.
* **Purpose:** Serves as the base DataFrame for calculating individual and cumulative percentage shares to evaluate categorical concentration.

In [72]:
complaint_dist = (
    df["Complaint_Type"].value_counts().reset_index(name="Ticket_Count")
)
complaint_dist

,Complaint_Type,Ticket_Count
0,Noise - Residential,30224
1,Heat/Hot Water,14711
2,Illegal Parking,9745
3,Blocked Driveway,3297
4,Street Condition,1310
...,...,...
128,Posting Advertisement,1
129,Pet Shop,1
130,Building Condition,1
131,Scaffold Safety,1


### Quantify Relative Volume Share Across Top 20 Complaint Types

Calculate the individual percentage share of total intake ($N = 85,122$) for each complaint category to assess workload distribution:

* **Tier 1 — Core Operational Drivers ($> 10\%$ each):**
  * `Noise - Residential`: **35.51%**
  * `Heat/Hot Water`: **17.28%**
  * `Illegal Parking`: **11.45%**
  * *Insight:* The top three categories alone account for **64.24%** of all resolved service requests, demonstrating that municipal intake is overwhelmingly dominated by quality-of-life and basic residential heating concerns.
* **Tier 2 — Moderate Recurring Volume ($1\% - 4\%$ each):**
  * Categories such as `Blocked Driveway` (**3.87%**), `Street Condition` (**1.54%**), `Unsanitary Condition` (**1.50%**), and `Abandoned Vehicle` (**1.40%**) capture consistent enforcement and infrastructure demands.
  * Commercial and street-level noise combine for an additional **2.48%** of ticket volume.
* **Tier 3 — Sub-1% Niche Categories:**
  * By Rank 14 (`Traffic Signal Condition` at **0.99%**), individual category contributions fall below $1.0\%$, highlighting a steep decay where the majority of NYC's 133 complaint types represent minor fractional shares of total city operations.

In [73]:
complaint_dist["Percentage"] = (complaint_dist["Ticket_Count"] / len(df) * 100).round(2)

In [74]:
complaint_dist[["Complaint_Type", "Percentage"]].head(20)

,Complaint_Type,Percentage
0,Noise - Residential,35.51
1,Heat/Hot Water,17.28
2,Illegal Parking,11.45
3,Blocked Driveway,3.87
4,Street Condition,1.54
5,Unsanitary Condition,1.50
6,Abandoned Vehicle,1.40
7,Noise - Commercial,1.24
8,Noise - Street/Sidewalk,1.24
9,Water System,1.12


### Evaluate Cumulative Volume Share Across Top 10 Complaint Types

Compute running cumulative percentages using `.cumsum()` to measure category concentration and assess the Pareto distribution of service requests:

* **Extreme Initial Concentration:**
  * **Top 2 Categories (52.79%):** `Noise - Residential` and `Heat/Hot Water` together comprise over half of all service request volume in the city.
  * **Top 3 Categories (64.24%):** Incorporating `Illegal Parking` brings the combined share to nearly two-thirds of all resolved complaints.
* **Top 10 Benchmark (76.15%):** Just 10 complaint types account for over three-quarters of the entire dataset ($64,821$ out of $85,122$ tickets).
* **Pareto Dynamics:** While 133 distinct complaint categories exist, operational intake exhibits extreme non-uniformity—less than $8\%$ of category types generate more than $76\%$ of all municipal workload, leaving the remaining 123 categories to share less than $24\%$.

In [ ]:
complaint_dist["Cumulative_Percentage"] = complaint_dist["Percentage"].cumsum().round(2)

complaint_dist.head(10)

,Complaint_Type,Ticket_Count,Percentage,Cumulative_Percentage
0,Noise - Residential,30224,35.51,35.51
1,Heat/Hot Water,14711,17.28,52.79
2,Illegal Parking,9745,11.45,64.24
3,Blocked Driveway,3297,3.87,68.11
4,Street Condition,1310,1.54,69.65
5,Unsanitary Condition,1273,1.50,71.15
6,Abandoned Vehicle,1195,1.40,72.55
7,Noise - Commercial,1058,1.24,73.79
8,Noise - Street/Sidewalk,1054,1.24,75.03
9,Water System,954,1.12,76.15


**Does volume matter?**

In [77]:
df.loc[df["Complaint_Type"] == "Uprooted Stump"]

,Unique_Key,Created_Date,Closed_Date,Agency,Complaint_Type,Descriptor,Borough,Incident_Zip,Status,Open_Data_Channel_Type,Resolution_Days
30697,63618380,2025-01-03 12:46:19,2025-02-27 10:36:51,DPR,Uprooted Stump,Remove Stump,Queens,11373.0,Closed,Phone,54
65805,63656468,2025-01-06 13:19:26,2025-10-31 08:53:01,DPR,Uprooted Stump,Remove Stump,Queens,11419.0,Closed,Phone,297


### Comparative Evaluation: Volume vs. Turnaround Across Fastest Categories

Inspect complaint volume (`count`) across service categories achieving same-day median resolution (`median == 0.0`):

* **High-Volume Velocity:** Rapid same-day resolution is maintained even under heavy operational load:
  * `Blocked Driveway`: **3,297 tickets** resolved with a median of **0.0 days**.
  * `Abandoned Vehicle`: **1,195 tickets** resolved with a median of **0.0 days**.
  * `Dirty Condition`: **932 tickets** resolved with a median of **0.0 days**.
* **Low-Volume Urgency:** Acute complaints with minimal occurrences—such as `Disorderly Youth` ($N = 1$), `Hazardous Materials` ($N = 21$), and `Asbestos` ($N = 17$) also resolve within 0 days due to immediate public safety routing.
* **Dual-Perspective Finding (Slowest vs. Fastest):**
  * The slowest complaints ($>60$ days) almost universally have tiny volumes ($N < 30$), ruling out system overload as the cause of delays.
  * The fastest complaints ($0.0$ days) encompass both single-ticket hazards and massive multi-thousand-ticket operations.
  * **Core Takeaway:** Resolution turnaround is driven by operational workflows (immediate field dispatch vs. contractor scheduling and administrative hearings), not incoming ticket volume.

In [83]:
is_volume_matter = (
    
    df.groupby("Complaint_Type")["Resolution_Days"]
    .agg(["median", "count"])
    .sort_values(by="median", ascending=False)

)

is_volume_matter.head(20)

,median,count
Complaint_Type,,
Uprooted Stump,175.5,2
New Tree Request,147.0,30
Noise - Helicopter,138.0,452
Special Projects Inspection Team (Spit),101.0,77
Green Taxi Complaint,96.5,4
Lot Condition,93.0,3
For Hire Vehicle Complaint,79.0,278
Taxi Complaint,76.5,124
Bench,69.5,2


In [84]:
is_volume_matter = (
    
    df.groupby("Complaint_Type")["Resolution_Days"]
    .agg(["median", "count"])
    .sort_values(by="median", ascending=True)

)

is_volume_matter.head(20)

,median,count
Complaint_Type,,
Abandoned Vehicle,0.0,1195
Animal In A Park,0.0,65
Asbestos,0.0,17
Animal-Abuse,0.0,207
Blocked Driveway,0.0,3297
Building Condition,0.0,1
Bike/Roller/Skate Chronic,0.0,6
Best/Site Safety,0.0,5
Derelict Vehicles,0.0,731
